draft/real-world-phoenix-lets-deploy-using-kubernetes.md

```diff
@@ -0,0 +1,643 @@
+
+If we want to automate any of this stuff we'll have to be able to talk to
+digitalocean programmatically, so installing doctl is step one. Please refer to
+this guide to get that setup: [doctl up and running
+guide](https://blog.digitalocean.com/introducing-doctl/)
+
+After you have `doctl` installed, make sure you create a personal access token
+in digitalocean and add that in an environment variable and also initialize your
+account.
+```bash
+export DIGITALOCEAN_TOKEN=[token]
+doctl auth init -t $DIGITALOCEAN_TOKEN
+doctl get account
+```
+
+### Install Terraform
+
+Ofcourse there is a handy install guide for terraform also. Actually just an
+executable, as it is written in go, so they have a handy packaged binary you can use.
+See: [terraform install](https://www.terraform.io/downloads.html) 
+
+### Setup our project
+
+If you want to follow along with this guide, create a folder that will hold your
+terraform configs and create a `main.tf` file that will hold our terraform declarations.
+
+```bash
+mkdir kube-terra
+cd kube-terra
+touch main.tf
+```
+
+### Setup the digitalocean Kubernetes cluster
+
+Terraform works in a declaritive manner, which means that we state what we want
+the world to be like and terraform figures out how to get to that state. In
+terraform you describe an item you want to exist as a resource. A Kubernetes
+cluster is a form of a resource that is provided by one of the many providers
+that exist in terraform. So let's create a digitalocean kubernetes cluster:
+
+Creating a resource is always in the format: `resource [kind] [reference] {}`
+We can use the reference we provide later on to refer back to this resource.
+
+```json
+resource "digitalocean_tag" "kubernetes-cl01" {
+  name = "kubernetes-cl01"
+}
+
+resource "digitalocean_kubernetes_cluster" "cl01" {
+  name    = "cl01"
+  region  = "ams3"
+  version = "1.16.2-do.1"
+
+  node_pool {
+    name       = "default"
+    size       = "s-1vcpu-2gb"
+    node_count = 2
+    tags       = ["${digitalocean_tag.kubernetes-cl01.id}"]
+  }
+}
+```
+
+Now you can get your cluster setup by issuing the following commands in the folder you
+created:
+
+```bash
+tarraform init
+terraform apply
+```
+
+This will take about 5 minutes and then you'll have a fresh kubernetes cluster
+up and running. Now that was easy!
+
+### Install Traefik
+
+Kubernetes needs an ingress controller to route traffik from the outside world
+to the services that are running inside the cluster as these services are not
+exposed to the outside world, which is a good thing! Nginx is often used, but I went with Traefik, a very nice alternative that has a lot of
+nice functinality out of the box, like automatic ssl certificates using Let's Encrypt
+and auto discovery of services running in the cluster.
+
+Create a file called `traefik-values.yml` in the root of your kubernetes config
+directory and add the following content. If you manage dns settings with
```

Comment:
FYI: using DNS as your acme provider also allows you to get a valid certificate when your production is behind a firewall.